### PR TITLE
Constructing Cliffords from quantum circuits with other Cliffords

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary (updated)

In #7966 we started adding `Cliffords` to quantum circuits _natively_, without explicitly converting them to `Instructions`. Unfortunately, this also broke some of the functionality for circuits containing Cliffords, see #9159. This PR fixes the problem of constructing a Clifford from a quantum circuit that contains other Cliffords. Moreover, the performance of doing so is greatly increased as the code now uses Clifford's `compose` function instead of decomposing the inner clifford into simple gates.

In addition, the `CollectCliffords` transpiler pass has been upgraded to collect and combine "clifford gates", where the "clifford gates" may now also include objects of type `LinearFunction`, `Clifford`, and `PauliGate`. 
_Question_: are there any other Clifford-like gates that can also be treated here?

### Details and comments

**Bug fix and performance improvement.** 

The code snippet 
```
qc = QuantumCircuit(1)
qc.append(qi.random_clifford(1), [0])
qi.Clifford(qc)
```
attempts to create a Clifford object from a quantum circuit that contains another clifford `cliff` (created by `random_clifford`). I have experimented with two possible fixes for this: converting `cliff` to circuit (as would have been done previously when converting `cliff` to instruction and using its `definition`) vs. directly using the Clifford's `compose` function (which is possible because `cliff` is of type `Clifford` and not `Instruction`). The second approach seems to be considerably faster, for example on my laptop the following code
```
    width = 20
    depth = 100
    rng = np.random.default_rng(12345)
    random.seed(1234)
    qc = QuantumCircuit(width)
    for _ in range(depth):
        w = random.randint(1, width)
        qargs = random.sample(range(width), w)
        cliff = random_clifford(w, seed=rng)
        qc.append(cliff, qargs)

    start_time = time.time()
    cliff = Clifford(qc)
    end_time = time.time()
    print(f"Time = {end_time - start_time:.4f}")
```
takes 4.15 seconds using `compose` and 25.88 seconds using `to_circuit`. This is not surprising, since Clifford synthesis is time-consuming for large cliffords.

**Update:** with #7483 merged (`Clifford.compose` becoming incredibly faster), the above code now takes 0.14 seconds!

**Extending the `CollectCliffords` transpiler pass**. 

This was very simple as the current code for `Clifford(circuit)` (in `clifford_circuits.py`) can already construct Cliffords from quantum circuit with linear functions and pauli gates.

### CODE REVIEW

**Comment by jakelishman** [comment]
Luciano: no, this is precisely against the data model.  Code that assumes that's the type of circuit and dag elements _shouldn't_, because it isn't any more.  That's the point of `Operation`.

File: test/python/transpiler/test_clifford_passes.py
Comment (ShellyGarion) [inline test/python/transpiler/test_clifford_passes.py:563]: Will this test also work? 

```
cliff1 = Clifford(qc)
cliff2 = Clifford(qc2)
self.assertTrue(cliff1 == cliff2)
```

The same question applies to the other tests as well.

Comment (alexanderivrii) [reply test/python/transpiler/test_clifford_passes.py:563]: Yes, it would, given the fix in `clifford_circuits.py` in this PR. Though please note that I have already added similar tests to `quantum_info/operators/symplectic/test_clifford`, i.e. that constructing a Clifford from quantum circuits (with other  Cliffords) produces the correct Clifford. Here in `transpiler/test_clifford_passes` I am more concerned with checking the correctness of the `CollectCliffords` transpiler pass (which under the hood does construct a Clifford from the given quantum circuit), so checking the equivalence of Operators seems slightly more appropriate.

**Comment by alexanderivrii** [comment]
One more comment. Once the PR #9157 gets merged, it would make sense to extend constructing Cliffords from quantum circuits with permutations.

File: test/python/transpiler/test_clifford_passes.py
Comment (ShellyGarion) [reply test/python/transpiler/test_clifford_passes.py:563]: I agree with this test, but will `test_collect_with_linear_functions` , `test_collect_with_pauli_gates` and `test_collect_with_all_types` will work with this code?

Comment (alexanderivrii) [reply test/python/transpiler/test_clifford_passes.py:563]: Thanks! So what was missing was a test in `quantum_info/operators/symplectic/test_clifford` that a Clifford can be constructed from a circuit that contains various clifford-like objects, that is regular clifford gates, linear functions, pauli gates, other cliffords, and other circuits with clifford-like objects. Added this test in [449cfcd](https://github.com/Qiskit/qiskit-terra/pull/9169/commits/449cfcda475ea2a88796345ee1d6966afd4cc023).

**Review by ShellyGarion — APPROVED** [review]
LGTM! Thanks!
We still need a code owner approval to get it merged.

File: qiskit/quantum_info/operators/symplectic/clifford_circuits.py
Comment (ikkoham) [inline qiskit/quantum_info/operators/symplectic/clifford_circuits.py:102]: Why is `copy` of `gate` necessary? (Sorry, I don't fully understand.)

**Review by ikkoham — COMMENTED** [review]
Thank you. Overall LGTM. I have one minor question.

File: qiskit/quantum_info/operators/symplectic/clifford_circuits.py
Comment (alexanderivrii) [reply qiskit/quantum_info/operators/symplectic/clifford_circuits.py]: Thanks! I have removed the unnecessary copying in [17f7ce0](https://github.com/Qiskit/qiskit-terra/pull/9169/commits/17f7ce0b8cecc69046951951ca749607f3dc88b6).

File: releasenotes/notes/improve-collect-cliffords-f57aeafe95460b18.yaml
Comment (jakelishman) [inline releasenotes/notes/improve-collect-cliffords-f57aeafe95460b18.yaml:7]: We usually use `upgrade` for API changes, where users will actively need to adjust their code to handle something we do differently now.  I may misunderstand a little, but I think this is more like a new feature - the `CollectCliffords` pass is now just more efficient, and can collect more objects into a single `Clifford`.

File: test/python/quantum_info/operators/symplectic/test_clifford.py
Comment (jakelishman) [inline test/python/quantum_info/operators/symplectic/test_clifford.py:487]: Does this produce a deterministic Clifford? If so, can we stick a test at the end to make sure it's producing the correct Clifford?

File: test/python/transpiler/test_clifford_passes.py
Comment (jakelishman) [inline test/python/transpiler/test_clifford_passes.py:563]: These tests use `equiv` rather than `==` - the former normalises global phase, so `==` is a strong condition.  Do we need the global-phase fuzziness here?  If so, it seems a bit unnerving, since the transpiler is supposed to maintain the global phase entirely.

(Comment repeats for all the tests here.)

**Review by jakelishman — COMMENTED** [review]
This looks pretty solid to me, thanks.  I just had a couple of minor points on the tests, but that's it.

For posterity: this PR isn't intended to fix all the issues raised by #9159, just a couple of them.  We're still working on all of #9159, but it will likely require a lot more overhauls to how inverses, controls and some other attributes are applied to gates.

File: test/python/quantum_info/operators/symplectic/test_clifford.py
Comment (jakelishman) [reply test/python/quantum_info/operators/symplectic/test_clifford.py]: Oh whoops, this comment must have been left in the review from weeks ago.

File: releasenotes/notes/improve-collect-cliffords-f57aeafe95460b18.yaml
Comment (alexanderivrii) [reply releasenotes/notes/improve-collect-cliffords-f57aeafe95460b18.yaml]: Thanks. I have indeed misunderstood what `update` means, fixed in [d9d9c17](https://github.com/Qiskit/qiskit-terra/pull/9169/commits/d9d9c178aaf952905e8ddcc0c5f511f785cb889e).

File: test/python/quantum_info/operators/symplectic/test_clifford.py
Comment (alexanderivrii) [reply test/python/quantum_info/operators/symplectic/test_clifford.py]: Good point, the resulting Clifford is deterministic, and I have expanded the test in [704d385](https://github.com/Qiskit/qiskit-terra/pull/9169/commits/704d3859443f4c4103544f17ff0d97788e69e3dd).

File: test/python/transpiler/test_clifford_passes.py
Comment (alexanderivrii) [reply test/python/transpiler/test_clifford_passes.py:563]: Ouch, your comment points to a real problem with `OptimizeCliffords` transpiler pass, that I should have but did not think about. 

The problem is that Cliffords are only defined up to a global phase: while "in the real world" `XY = iZ`, when using Clifford's stabilizer/destabilizer tableaus, the multiplication by `i` get dropped, i.e. "in the Clifford world" `XY=Z`. This is also the reason why most of the tests that construct an `Operator` out of `Clifford` in `test_clifford.py` check for `equiv` and not for `==`. 

The above in code:
```
qc = QuantumCircuit(1)
qc.x(0)
qc.y(0)
print(Operator(Clifford(qc)) == Operator(qc))
```
print the value `False`.

So right now replacing a block of Clifford gates by a Clifford may drop the global phase, which is indeed the problem to "the transpiler maintaining the global phase". That is, the `OptimizeCliffords` transpiler pass can only guarantee equivalence up to a global phase.

Note: I do believe that when optimizing Cliffords over _a subset of qubits_ the problem is still only with _global_ and not _local_ phases.

BTW, I thought there was an effort to incorporate the full phase into the `Clifford` class, maybe @ShellyGarion or @ikkoham can comment? This would immediately solve our problems here.

Comment (alexanderivrii) [reply test/python/transpiler/test_clifford_passes.py:563]: So I don't think we should replace `equiv` by `==` in these tests. However, we should see if there is already a solution for maintaining the global phase. And if not we should clearly document this caveat in the documentation for `OptimizeCliffords` transpiler pass. What do you think?

Comment (jakelishman) [reply test/python/transpiler/test_clifford_passes.py:563]: That sounds like a sensible plan to me. If the phase non-equivalence is already in the pass and in Terra 0.22, then there's no massive rush to fix it now before the release, and anyway, we can treat it as a bug fix. Let's open a new issue and discuss the right way forwards.

**Review by jakelishman — APPROVED** [review]
Potential global-phase issues aside, which are for another issue, this looks good to merge to me.

**Comment by jakelishman** [comment]
@Mergifyio requeue

**Comment by jakelishman** [comment]
@Mergifyio requeue

**Comment by mtreinish** [comment]
@Mergifyio unqueue

**Comment by mtreinish** [comment]
@Mergifyio requeue

**Pull request merged**
